nexus/types/src/external_api/params.rs

```diff
@@ -83,6 +89,44 @@ pub struct OptionalProjectSelector {
     pub project_selector: Option<ProjectSelector>,
 }
 
+#[derive(Deserialize, JsonSchema)]
+pub struct DiskSelector {
+    #[serde(flatten)]
+    pub project_selector: Option<ProjectSelector>,
+    pub disk: NameOrId,
+}
+
+impl DiskSelector {
+    pub fn new(
+        organization: Option<NameOrId>,
+        project: Option<NameOrId>,
+        disk: NameOrId,
+    ) -> Self {
+        DiskSelector {
+            project_selector: project
+                .map(|p| ProjectSelector::new(organization, p)),
+            disk,
+        }
+    }
+}
+
+#[derive(Deserialize, JsonSchema)]
+pub struct DiskList {
+    #[serde(flatten)]
+    pub project_selector: ProjectSelector,
+    #[serde(flatten)]
+    pub pagination: PaginatedByName,
+}
+
+#[derive(Deserialize, JsonSchema)]
+pub struct DiskMetricsList {
+    #[serde(flatten)]
+    pub pagination: PaginationParams<ResourceMetrics, ResourceMetrics>,
```

Comment:
I don't understand how this works. I would expect that the second type parameter to `PaginationParams` would need to have some way to encode `disk_selector` in `disk_metrics_list_v1` or else subsequent iterations would need to continue to pass in the properties of `ProjectSelector` (FYI, progenitor assumes that this is not necessary).

I'd suggest

```suggestion
pub struct DiskMetricsListParams {
    #[serde(flatten)]
    pub metrics: ResourceMetrics,
    
    #[serde(flatten)]
    pub project_selector: Option<ProjectSelector>,
}

pub struct DiskMetricsListSelector {
    pub metrics: ResourceMetrics,
    pub disk_select: DiskSelector,
}
```

Then `disk_metrics_list_v1` would do `query_params: Query<PaginationParams<params::DiskMetricsListParams, params::DiskMetricsListSelector>>,`

Reply:
I'd also note, that I think this would be easier to understand if it lived with the endpoint. The organization of endpoint/param/view means that related items are not contiguous.

Reply:
Good catch. Yeah, this is wrong. Let me do some house keeping.